src/lib/builders/color-picker/create.ts

```diff
@@ -0,0 +1,898 @@
+import { addMeltEventListener, builder, createElHelpers, executeCallbacks, isBrowser } from "$lib/internal/helpers";
+import type { Defaults } from "$lib/internal/types";
+import { onMount } from "svelte";
+
+import type { ArrowKeys, ColorHSL, ColorHSV, ColorPickerParts, ColorRGB, CreateColorPickerProps, EyeDropperType, EyeDropperWindow, KeyDurations, NodeElement, NodeSize, Position, ReturnedColor } from "./types";
+import { derived, get, writable, type Readable, type Writable } from "svelte/store";
+
+
+const defaults = {
+    forceVisible: false,
+    defaultColor: '#ff0000',
+    hueSliderOrientation: 'horizontal',
+    alphaSliderOrientation: 'horizontal',
+} satisfies Defaults<CreateColorPickerProps>;
+
+const { name } = createElHelpers<ColorPickerParts>('color-picker');
+
+
+export function createColorPicker(args?: CreateColorPickerProps) {
+    const argsWithDefaults = { ...defaults, ...args };
+
+    let dragging = false;
+    let hueDragging = false;
+    let alphaDragging = false;
+    let insideUpdate = false;
+    let inputUpdate = false;
+    const speepUpStep = 5;
+
+    const keyDurations: KeyDurations = {
+        'ArrowDown': null,
+        'ArrowLeft': null,
+        'ArrowUp': null,
+        'ArrowRight': null
+    };
+
+    const color: Writable<string> = writable(defaults.defaultColor);
+
+    const colorCanvasDims: Writable<NodeElement<HTMLCanvasElement>> = writable({ height: 1, width: 1 });
+    const colorPickerDims: Writable<NodeSize> = writable({ height: 1, width: 1 });
+    const colorPickerPos: Writable<{ x: number; y: number; }> = writable({ x: 0, y: 0 });
+
+    const hueSliderDims: Writable<NodeElement<HTMLCanvasElement>> = writable({ height: 1, width: 1 });
+    const huePickerDims: Writable<NodeSize> = writable({ height: 1, width: 1 });
+    // const huePickerPos: Writable<Position> = writable({ x: 0, y: 0 });
+    const hueAngle: Writable<number> = writable(0);
+
+    const alphaSliderDims: Writable<NodeElement<HTMLCanvasElement>> = writable({ height: 1, width: 1 });
+    const alphaPickerDims: Writable<NodeSize> = writable({ height: 1, width: 1 });
+    const alphaValue: Writable<number> = writable(100);
+
+    let eye: EyeDropperType | null = null;
+
+    const isValidHex = (hex: string) => /^#([0-9a-f]{3}){1,2}$/i.test(hex);
+
+    const setAlphaValue = (x: number) => alphaValue.set(Math.round(x / get(alphaSliderDims).width * 100));
+
+    /**
+     * Update the canvases when the color changes.
+     */
+    // const updateCanvases = derived([color], ([$color]) => {
+    //     if (isValidHex($color)) updateOnColorInput($color);
+    // });
+
+    /**
+     * Returns the current color of the color picker.
+     */
+    const getCurrentColor: Readable<() => ReturnedColor> = derived([hueAngle, colorPickerPos, colorCanvasDims, alphaValue], ([$hueAngle, $colorPickerPos, $colorCanvasDims, $alphaValue]) => {
```

Comment:
Why is this needed when you already have `$color`? 

If it's to get access to RGB values, maybe it makes more sense to have derived states, one for RGB, one for HSL, on for HEX, and one for HSV (am I missing more?). Or even just a single derived state with all of them. And it doesn't need to be a function I believe

Reply:
@TGlide It's actually a leftover from a previous implementation where I was using a different approach. I think having one `$value` store with a hex value, and a derived store with different color representations makes sense. I'll add it.